util/dvsim/FpvCfg.py

```diff
@@ -0,0 +1,253 @@
+                    email_result["pass_rate"],
+                    email_result["cov_rate"]
+                ])
+            messages = cfg.result.get("messages")
+            if messages:
+                error = self.parse_dict_to_str(messages)
+                if error:
+                    error_message += "\n####" + cfg.name + "\n"
+                    error_message += error
+
+        if len(email_table) > 1:
+            self.email_summary_md = results_str + tabulate(
+                email_table, headers="firstrow", tablefmt="pipe", colalign=colalign)
+        self.email_summary_md += error_message
+
+        return self.results_summary_md
+
+    def _gen_results(self):
+        # '''
+        # This function is called after the regression and looks for
+        # results.hjson file with aggregated results from the FPV run.
+        # The hjson file is required to follow this format:
+        # {
+        #   "messages": {
+        #      "errors"      : []
+        #      "warnings"    : []
+        #      "cex"         : ["property1", "property2"...],
+        #      "undetermined": [],
+        #      "unreachable" : [],
+        #   },
+        #
+        #   "fpv_summary": {
+        #      "proven"      : 20,
+        #      "cex"         : 5,
+        #      "covered"     : 18,
+        #      "undetermined": 7,
+        #      "unreachable" : 2,
+        #      "pass_rate"   : "90 %",
+        #      "cover_rate"  : "90 %"
+        #   },
+        # }
+        # The categories for property results are: proven, cex, undetermined,
+        # covered, and unreachable.
+        #
+        # If coverage was enabled then results.hjson will also have an item that
+        # shows FPV coverage. It will have the following format:
+        #   "fpv_coverage": {
+        #      Stimuli: "90 %",
+        #      COI:     "90 %",
+        #      Proof:   "80 %"
+        #   }
+        # '''
+        results_str = "## " + self.results_title + "\n\n"
+        results_str += "### " + self.timestamp_long + "\n"
+        results_str += "### FPV Tool: " + self.tool.upper() + "\n"
+        results_str += "### LogFile dir: " + self.scratch_path + "/default\n\n"
+
+        colalign = ("center", ) * len(self.header)
+        table = [self.header]
+        summary = [self.name]  # cfg summary for publish results
+
+        for mode in self.build_modes:
+            result_data = Path(subst_wildcards(self.build_dir, {"build_mode": mode.name}) +
+                               '/results.hjson')
+            try:
+                with open(result_data, "r") as results_file:
+                    self.result = hjson.load(results_file, use_decimal=True)
+            except IOError as err:
+                log.warning("%s", err)
+                self.result = {
+                    "messages": {
+                        "errors": ["IOError: %s" % err],
+                    }
+                }
+
+            fpv_summary = self.result.get("fpv_summary")
```

Comment:
I'd consider pulling the body of the for loop from here into a separate function (just because it's pretty long). Or maybe make two functions: one for dealing with `fpv_summary` and one for dealing with `fpv_coverage`. Either way, `_gen_results` is a bit long at the moment.

Also I'd probably pass the loaded hjson data as an argument, rather than using `self.result` as a "global variable", which is rather ugly. (I don't think it's used anywhere else, and you're overwriting it on each loop iteration, which is a sign that something odd is going on).

Reply:
Ah, I see. That does mean that you only report results from the last build mode. Is that intentional?

Reply:
sorry I forget to address this one, that is a good point. Currently we only support one "default mode" in FPV, so this `build_modes` should only have one item in the list. I will add some logic to clarify that.